### ISSUE TITLE
Feature Request: Enable DAPR integration in Container Jobs

### ISSUE THREAD

**Issue by penfold** [microsoft/azure-container-apps#1000]
**Is your feature request related to a problem? Please describe**.  
I want to be able to invoke actions using DAPR services discovery or pubsub from a container running as a job in a Container App Environment.

**Describe the solution you'd like**.    
The DAPR sidecar is not loaded with Container jobs and there is no config for DAPR in bicep specs. 

**Describe alternatives you've considered**.   
I could run the container using DAPR cron but I want the option to run it manually too, so using jobs feels like the correct approach.

**Additional context**.

**Comment by berndverst** [comment]
Dapr Maintainer here - there are some major caveats when running Dapr with Kubernetes jobs. See https://docs.dapr.io/operations/hosting/kubernetes/kubernetes-job/.

We will have to weigh these caveats to see if we can find a good solution around them. If unresolved providing this functionality would be more trouble than it is worth (high likelihood that jobs would never terminate and hence incur compute costs).

**Comment by penfold** [comment]
It sounds like there needs to be then possibility to call into Kubernetes to say 'Job completed - don't restart'. This with a call to DAPR shutdown (or as part of DAPR) would remove the restarted issues.

**Comment by berndverst** [comment]
> It sounds like there needs to be the possibility to call into Kubernetes to say 'Job completed - don't restart'. This with a call to DAPR shutdown (or as part of DAPR) would remove the restarted issues.

Yes, and then you need to figure out a way to do so for the many users who will forget to call shutdown and then open support tickets about their "unexpected bills" :) That's the real problem.

*If* the product team decides to prioritize this we would have to follow a very different approach here to decouple the Dapr sidecar from the Job lifecycle.

**Comment by Happycoil** [comment]
I use ACA Jobs extensively, so I'm very interested in this feature.

@berndverst the new [native sidecar support](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) in k8s should do most of the heavy lifting to resolve the dapr lifecycle issue in jobs, right? Is getting this feature into ACA on the roadmap?

**Comment by Steve887** [comment]
Has there been any movement on this in the last 6 months?

**Comment by nilshjalmarson** [comment]
I am also interssted in this. 

My use case is this:
I have a job that will download data from an external service once a day. A job is good for this with the native sceduling.
But then I whant to post a message on a topic so that a number of other apps can process the data and send it to different locations.
I whanted to use Dapr to send the message since it will simplify things and then I could use Aspire to set up the whole flow locally in a very simple way.

**Comment by BlackRider97** [comment]
This remains a valid feature request. Thank you for the continued interest and use case descriptions.

The key technical challenge is Dapr sidecar lifecycle management in Jobs — the sidecar can prevent job termination, leading to unexpected compute costs. While [Kubernetes native sidecar support](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) is now available in ACA, integrating it with the Jobs model requires additional design work that we haven't prioritized yet.

For job workloads that need Dapr capabilities today, consider these workarounds:
1. **KEDA-scaled Container App** (scale to zero when idle) instead of a Job
2. **HTTP calls from the Job** to a Dapr-enabled Container App for pubsub/state operations

We don't have a timeline to share for this feature at this point, but we appreciate the feedback and will keep this issue open for tracking community interest.

---
*🤖 This comment was generated with AI assistance and may contain inaccuracies. If this doesn't address your concern or if the issue persists, please reopen this issue or create a new one.*
